.. inStrain documentation master file, created by
   sphinx-quickstart on Sat Sep 21 13:16:05 2019.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

inStrain
====================================


inStrain is a program for measuring, comparing, and interrogating the genetic heterogeneity of microbial populations in and between metagenomic samples. This includes calculation of nucleotide diversity, calling SNPs (including non-synonymous and synonymous variants), reporting accurate coverage / breadth, and calculating linkage disequilibrium in the contexts of genomes, contigs, and individual genes.
Source code is `available on GitHub
<https://github.com/MrOlm/instrain>`_.

See links to the left for :doc:`installation` and :doc:`quickstart` instructions

Comments and suggestions can be sent to `Matt Olm <mattolm@berkeley.edu>`_ and/or
`Alex Crits-Christoph <crits-christoph@berkeley.edu>`_

Bugs reports and feature requests can be submitted through `GitHub
<https://github.com/MrOlm/instrain/issues>`_.

Contents
--------

.. toctree::
   :maxdepth: 2

   overview
   installation
   quickstart
   preparing_input
   module_descriptions
   choosing_parameters
   example_output
   Advanced_use
