.. inStrain documentation master file, created by
   sphinx-quickstart on Sat Sep 21 13:16:05 2019.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

InStrain
====================================

InStrain is a tool for  analysis of co-occurring genome populations from metagenomes that allows
highly accurate genome comparisons, analysis of coverage, microdiversity, and linkage,
and sensitive SNP detection with gene localization and synonymous non-synonymous identification

Source code is `available on GitHub
<https://github.com/MrOlm/instrain>`_.

See links to the left for :doc:`installation` instructions

Comments and suggestions can be sent to `Matt Olm <mattolm@berkeley.edu>`_ and/or
`Alex Crits-Christoph <crits-christoph@berkeley.edu>`_

Bugs reports and feature requests can be submitted through `GitHub
<https://github.com/MrOlm/instrain/issues>`_.

Contents
--------

.. toctree::
   :maxdepth: 2

   installation
   overview
   program_documentation
   example_output
   tutorial
   Advanced_use
