.. inStrain documentation master file, created by
   sphinx-quickstart on Sat Sep 21 13:16:05 2019.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

InStrain
====================================

InStrain is a tool for  analysis of co-occurring genome populations from metagenomes that allows highly accurate genome comparisons, analysis of coverage, microdiversity, and linkage,
and sensitive SNP detection with gene localization and synonymous non-synonymous identification

Source code is `available on GitHub
<https://github.com/MrOlm/instrain>`_.

Publication is available in `Nature Biotechnology <https://doi.org/10.1038/s41587-020-00797-0>`_ and on `bioRxiv <https://www.biorxiv.org/content/10.1101/2020.01.22.915579v1>`_

See links to the left for :doc:`installation` instructions

Bugs reports and feature requests can be submitted through `GitHub
<https://github.com/MrOlm/instrain/issues>`_.

InStrain was developed by `Matt Olm <mattolm@stanford.edu>`_ and
`Alex Crits-Christoph <crits-christoph@berkeley.edu>`_ in the `Banfield Lab <https://geomicrobiology.berkeley.edu/>`_ at the University of California, Berkeley.

Contents
--------

.. toctree::
   :maxdepth: 2

   installation
   overview
   important_concepts
   tutorial
   user_manual
   example_output
   advanced_use
   benchmarks
   acknowledgements
